### ISSUE TITLE
Fetching from Facebook

### ISSUE THREAD

**Issue by CoolestNerdIII** [TID-Lab/aggie#139]
When trying to fetch from Facebook sources, I have received the following error:
```
(#12) fql is deprecated for versions v2.1 and higher
```
The steps required to create a new application on Facebook and request and API were followed. Any help would be greatly appreciated.

**Comment by hooverlunch** [comment]
What URL are you using? It should still work until late 2016...

**Comment by CoolestNerdIII** [comment]
I used the URL 
https://graph.facebook.com/oauth/access_token?client_secret=xxx&client_id=xxx&grant_type=client_credentials

Replacing the first xxx with the App Secret 
Replacing the second xxx with the App ID

**Comment by CoolestNerdIII** [comment]
I also tried changing the fql version, 

https://graph.facebook.com/v2.1/oauth/access_token?client_secret=f574eb47a1b0ca40d630694ac2acdaab&client_id=1590455831200410&grant_type=client_credentials

**Comment by CoolestNerdIII** [comment]
I also tried changing the FQL version as suggested by StackOverflow by using:
````
https://graph.facebook.com/v2.0/oauth/access_token?client_secret=xxx&client_id=xxx&grant_type=client_credentials

```
And using different versions, but also to no avail.

**Comment by hooverlunch** [comment]
You shouldn't have to put those URLs into Aggie. Just put the URL of the group or page you want to track in the source dialog and it should just work. The interaction with the FB API is handled by Aggie.

**Comment by CoolestNerdIII** [comment]
I'm not sure I understand your response. 
From following the instructions, I:
- Created an Application on Facebook
- From the FaceBook application, I retrieved the App ID and the App Secret
- In the URL, I used the link mentioned above, replacing the 'xxx' with the corresponding app id and app secret key
- I visited the URL and retrieved an access token
- I placed the access token in the 'secrets.json' file and then began running aggie
- Within Aggie, when adding a source, I added a FaceBook group
- Once I began fetching, the error mentioned above occurred. 

Is there a step that is missing?

**Comment by hooverlunch** [comment]
Hmm, what version of the fbgraph package do you have?

On 31 March 2015 at 19:11, Paul Wilson <notifications@github.com> wrote:

> I'm not sure I understand your response.
> From following the instructions, I:
>
>    - Created an Application on Facebook
>    - From the FaceBook application, I retrieved the App ID and the App
>    Secret
>    - In the URL, I used the link mentioned above, replacing the 'xxx'
>    with the corresponding app id and app secret key
>    - I visited the URL and retrieved an access token
>    - I placed the access token in the 'secrets.json' file and then began
>    running aggie
>    - Within Aggie, when adding a source, I added a FaceBook group
>    - Once I began fetching, the error mentioned above occurred.
>
> Is there a step that is missing?
>
> --
> Reply to this email directly or view it on GitHub
> <https://github.com/TID-Lab/aggie/issues/139#issuecomment-88279521>.
>



-- 
Tom Smyth

Worker-Owner, Sassafras Tech Collective
Specializing in innovative, usable tech for social change
sassafras.coop *·* @sassafrastech

Resident, Touchstone Cohousing
touchstonecohousing.org

**Comment by CoolestNerdIII** [comment]
fbgraph@0.2.13

**Comment by hooverlunch** [comment]
Hmm, it could be that you might have to adjust the API version in your
Facebook app. I'm not sure.

The fbgraph package is not very complicated. You could dive in there and
see where the API version is getting specified.

The key think will be to make sure version 2.0 of the FB API is getting
used. That's the last version that supports FQL.

On 31 March 2015 at 20:23, Paul Wilson <notifications@github.com> wrote:

> fbgraph@0.2.13
>
> --
> Reply to this email directly or view it on GitHub
> <https://github.com/TID-Lab/aggie/issues/139#issuecomment-88292378>.
>



-- 
Tom Smyth

Worker-Owner, Sassafras Tech Collective
Specializing in innovative, usable tech for social change
sassafras.coop *·* @sassafrastech

Resident, Touchstone Cohousing
touchstonecohousing.org

**Comment by CoolestNerdIII** [comment]
Ok I will look into it and report here with anything/everything. 

Thanks for your help!

**Comment by CoolestNerdIII** [comment]
@hooverlunch What is your current version of fbgraph?

**Comment by hooverlunch** [comment]
"fbgraph": "~0.2.9"

On 7 April 2015 at 16:52, Paul Wilson <notifications@github.com> wrote:

> @hooverlunch <https://github.com/hooverlunch> What is your current
> version of fbgraph?
>
> --
> Reply to this email directly or view it on GitHub
> <https://github.com/TID-Lab/aggie/issues/139#issuecomment-90729345>.
>



-- 
Tom Smyth

Worker-Owner, Sassafras Tech Collective
Specializing in innovative, usable tech for social change
sassafras.coop *·* @sassafrastech

Resident, Touchstone Cohousing
touchstonecohousing.org

**Comment by CoolestNerdIII** [comment]
So we have I believe come to a conclusion about the error located [here](http://stackoverflow.com/questions/25256428/query-facebook-for-what-version-of-the-graph-api-is-being-used-can-be-used) in stackoverflow. 

Basically, if an Facebook App was created post August 7, 2014, there is no way to revert to back to API version 2.0. The application that we have created on Facebook is relatively new, and after trying a variety of different things, I am unsure if there are any work arounds that could resolve the problem.

What should be our next approach? 
Would it be:
- Is it possible that the facebook functions would be rewritten relatively soon?
- Should we begin the process of rewriting 'facebook-content-service.js' in effort to replace fql, and if so, could we possibly receive guidance on this approach?
- Are you aware of any hacks that could possibly resolve the problem?

Thanks!

**Comment by hooverlunch** [comment]
I feared it might be something like that. Bummer. I am not aware of any
hacks.

I think the best thing to do is just to bite the bullet and rewrite. As of
now Sassafras has no funding for Aggie development and we are occupied with
other things so I'd suggest giving it a try yourselves.

I would be happy to provide some pointers. The steps would be as follows,
IMO:

1. Review the tests in
test/lib.fetching.content-services.facebook-content-service.test.
1. Identify and experiment with the Graph API methods that can get the
needed data. Given the nature of the Graph API, I feel like you will end up
needing multiple requests to it to get what we were getting in one request
to FQL.
1. As such, I would create multiple methods in the class: _getPosts,
_getComments, and _getAuthors. This will allow you to stub each one
separately in the tests. You can just leave them blank at first.
1. Update the Facebook fixture files in test/fixtures to match the data
that would be returned from each call to each of those functions for each
of the existing test cases. You'll want to reuse the same data (basically
transposing the FQL results to Graph API results) so that the test
assertions won't need to be changed. Be sure that the new fixture data
matches exactly what would be returned from Facebook.
1. Update the stubWithFixture method to perhaps take a filename prefix
instead of a full filename, and then make it stub each of the three methods
with fixtures based on that prefix.
1. Run the tests and watch them fail.
1. Write the code to make them pass.

I would suggest starting with one test case (perhaps the empty one) first
and working through these steps. You can run one mocha test at a time (I
forget the command syntax).

If you want to submit pull requests as you go I can try to provide more
suggestions.

Thanks for taking this on!



On 7 April 2015 at 18:32, Paul Wilson <notifications@github.com> wrote:

> So we have I believe come to a conclusion about the error located here
> <http://stackoverflow.com/questions/25256428/query-facebook-for-what-version-of-the-graph-api-is-being-used-can-be-used>
> in stackoverflow.
>
> Basically, if an Facebook App was created post August 7, 2014, there is no
> way to revert to back to API version 2.0. The application that we have
> created on Facebook is relatively new, and after trying a variety of
> different things, I am unsure if there are any work arounds that could
> resolve the problem.
>
> What should be our next approach?
> Would it be:
>
>    - Is it possible that the facebook functions would be rewritten
>    relatively soon?
>    - Should we begin the process of rewriting
>    'facebook-content-service.js' in effort to replace fql, and if so, could we
>    possibly receive guidance on this approach?
>    - Are you aware of any hacks that could possibly resolve the problem?
>
> Thanks!
>
> --
> Reply to this email directly or view it on GitHub
> <https://github.com/TID-Lab/aggie/issues/139#issuecomment-90749948>.
>



-- 
Tom Smyth

Worker-Owner, Sassafras Tech Collective
Specializing in innovative, usable tech for social change
sassafras.coop *·* @sassafrastech

Resident, Touchstone Cohousing
touchstonecohousing.org

**Comment by CoolestNerdIII** [comment]
Ok thanks and I will be in contact for any support :)

**Comment by pgarrison** [comment]
Fixed with #143
